Shared private data (#51)
wnfs-wg/spec: spec/shared-private-data.md

# WNFS Shared Private Data Extension Specification

# 0 Abstract

There are several cases where we need to exchange private data asynchronously when the other party may be offline.
The shared private data extension allows a file system's owner to deposit asymmetrically encrypted messages so they can be picked up by the recipient when they come online.
A recipient will then periodically - or after being prompted - scan the sender's private forest for payloads they can decrypt.
These encrypted payloads contain secrets giving read access and/or [UCANs](https://ucan.xyz) giving write access to parts of the owner's public or private file system.

# 1 Exchange Keys

To share information with a user that's offline we make use of asymmetric encryption. All WNFS users widely distribute a list of 2048-bit RSA public keys - their non-exportable "exchange keys" - as `did:key:` DIDs at a well-known location: A list under `exchange` at the root of WNFS next to `public` and `private`. See [the rationale](/rationale/shared-private-data.md#exchange-key-location) for more information.

> These RSA keys are used to encrypt a share payload for a recipient. This share payload contains a pointer to a private node and the symmetric key to decrypt it.

Reviewer: It wasn't initially clear to me that "share payload" formed a noun phrase, rather than a typo. Perhaps it'd help if you linked it to the section: https://github.com/wnfs-wg/spec/blob/matheus23/shared/spec/shared-private-data.md#22-share-payload

Allowing multiple exchange keys per WNFS enables multiple recipient devices to receive a share without having to transfer exchange keys between devices.

# 2 Data Layout

Shared private data payloads are stored in the [Private Forest](/spec/private-wnfs.md#21-ciphertext-blocks).

## 2.1 Share Label

> Shares are labeled in the private forest by a namefilter consisting of:

Reviewer: I'd also consider linking to the [namefilter spec](https://github.com/wnfs-wg/spec/blob/matheus23/shared/spec/namefilter.md) here :)

# WNFS Shared Private Data Extension Specification

# 0 Abstract

> There are several cases where we need to exchange private data asynchronously when the other party may be offline.

Reviewer 2: Can we list some example cases? If not here, in the rationale?

Author: Hm. Examples would include stuff like "device owner is asleep", "device is turned off" or "device is not connected to the internet".

I think perhaps instead I can motivate this via store-and-forward networks?

The shared private data extension allows a file system's owner to deposit asymmetrically encrypted messages so they can be picked up by the recipient when they come online.

> A recipient will then periodically - or after being prompted - scan the sender's private forest for payloads they can decrypt.

Reviewer 2: Would like recommendation(s) on how this can be done periodically or after being prompted?

These encrypted payloads contain secrets giving read access and/or [UCANs](https://ucan.xyz) giving write access to parts of the owner's public or private file system.

# 1 Protocol Versions

> The shared private data extension is versioned and designed to support multiple versions living on the same file system concurrently, so devices can gradually upgrade to the new version and support multiple versions at the same time.

Reviewer 3: Can you define "extension" in the context of this spec?

Author: I'm thinking of it like this: The rest of the spec doesn't have a dependency on shared private data existing. It's technically possible to ship an implementation of WNFS without shared private data and it'd work and be useful.
Not super attached to this idea, it was basically a "why not?".
Should I add a paragraph explaining this? Or should I rather just make shared private data part of the WNFS spec?

Reviewer 3: Ah, I see. Arguably the same can be said for private and public data not having dependencies on each other. If we want to go down the path of something being totally optional, it probably should live in its own spec with WNFS as a dependency. I would be surprised if someone _didn't_ want to implement this as a feature in an application, though at the spec level there's definitely an argument to be made.

The protocol version determines the algorithm used, key size and other parameters.

## 1.1 Version 1

Version 1 of the shared private data protocol uses RSA public keys for asymmetric encryption. These RSA public keys MUST have an exponent of 65537 and MUST be 2048-bit.
The method for encryption MUST be RSAES-OAEP.

# 2 Exchange Keys Partition

To share information with a user who is offline we make use of asymmetric encryption. All WNFS users widely distribute a list of public keys - their non-exportable "exchange public keys" at a well-known location: A map under `exchange` at the root of WNFS next to `public` and `private`. See [the rationale](/rationale/shared-private-data.md#exchange-key-location) for more information.

These keys are used to encrypt a [share payload](#32-share-payload) for a recipient. This share payload contains a pointer to a private node and the symmetric key to decrypt it.

> Allowing multiple exchange keys per WNFS enables multiple recipient devices to receive a share without having to transfer exchange keys between devices.

Reviewer 3: > without having to transfer exchange keys between devices.

It may be worth adding a sentence or two about why this is important

## 2.1 Layout

> The exchange key partition is a link to a [public directory](/spec/public-wnfs.md). This directory contains a set of further directories with exchange keys within.

Reviewer 3: Not sure if it's just me, but this took me a couple times to realize that it didn't mean that you're linking to the public partition

```suggestion
The exchange key partition is structured as a [public directory](/spec/public-wnfs.md) under its own top-level link. This directory contains further directories containing exchange keys.
```

Author: Yeah! Thanks, that's better.

> Exchange keys are grouped by device, such that a sender only needs to choose *one* of the available keys for each device. For example, senders that support protocol version 1 and 2 look through all recipient's devices and select either a single version 2 exchange key or, if no such key is present, a version 1 key from each device.

Reviewer 3: Why have it broken down by device rather than a list of public keys per version? So that they can be revoked by context later as needed?

I would also perhaps change the term from "device" to "agent" or "content". e.g. laptop Chrome vs laptop Firefox vs native.

Reviewer 3: (I should have highlighted lines 33-38 for the above)

Reviewer 3: From later, this is broken down as:

```
/exchange
├── aNXmyZ-kRI1-fsLcOol8wi0wfmxxY5_15bxdt2T4bs8
│   ├── v1.exchangekey
│   ├── v2.exchangekey
├── BEubij6fgCV51wyhYtITYCYhoGvk5Y5Slr59Kl4FOkQ
│   └── v1.exchangekey
└── zulwKufD7v71rJ7792cl2pi1LHmePS7d7a0vhuvaKNo
    └── v2.exchangekey
```

...why not...

```
/exchange
├── v1
│   ├── publicKey1
│   └── publicKey2
└── v2
    ├── publicKey3
    └── publicKey4

```

...or...

```
/exchange
├── v1.json
└── v2.json
```

...where the values of the `.json` is dag-json broken out into separate blocks (and thus can easily be checked to be append-only).

Author: > Why have it broken down by device rather than a list of public keys per version?

Because when creating share payloads, you want to pick *one key for each device*.
If you structure it differently, you lose the mapping of which groups of keys belong to one device. Then you're forced to encrypt for *every* exchange key, and since some devices advertise both v1 and v2 keys, you'll do some redundant work.

I tried to explain that reasoning with the paragraph in line 33.

See the [example exchange partition layout](#exchange-partition-layout) for examples.

### 2.1.1 Key Encoding

Individual device exchange keys are versioned. The protocol version is derived from the file name. Protocol version 1 exchange keys are named `v1.exchangekey`.

The file's content is 256 bytes of the exchange key's RSA public modulus encoded as low-endian unsigned integer.

See the [test vectors](#exchange-key-test-vectors) for examples.

### 2.2 Merging

The exchange partition is merged according to the rules for [public directory merging](/spec/public-wnfs.md#merge).

# 3 Shared Private Data

Shared private data payloads are stored in the [Private Forest](/spec/private-wnfs.md#21-ciphertext-blocks).

## 3.1 Share Label

Shares are labeled in the private forest by a [namefilter](/spec/namefilter.md) consisting of:
- The sender's root DID. In most cases this will be the file system owner's root DID.
- The recipient's exchange key
- A counter, which is incremented each time the sender creates a share.

The counter MUST be a 0-based natural number.

> The namefilter for a share is computed like this:

Reviewer 3: If following RFC 2119:

```suggestion
The namefilter for a share MUST be computed as follows:
```

```ts
function computeShareLabel(senderRootDID: string, recipientExchangeKey: ByteString, counter: number) {
  return new Namefilter()
    .add(senderRootDID)
    .add(recipientExchangeKey)
    .add(encode(counter))
    .saturate()
}
```

Here `encode` is a function that maps a block index to a low-endian byte array encoding of a 64-bit unsigned integer.

The `recipientExchangeKey` are the recipient device's exchange key bytes, including the protocol versioning prefix.

## 3.2. Share Payload

The share payload MUST be a non-empty list of CIDs to [RSAES-OAEP](https://datatracker.ietf.org/doc/html/rfc3447#section-7.1)-encrypted ciphertexts.

> When decrypted, this reveals a payload of the form:
> 
> `<unsigned varint prefix><32 byte private forest label><32 byte AES key>`
> 
> This is a [multicodec](https://github.com/multiformats/multicodec/).

Reviewer 3: ```suggestion
The decrypted payload MUST be represented as a [multicodec](https://github.com/multiformats/multicodec/) in the form:

`<unsigned varint prefix><32 byte private forest label><32 byte AES key>`
```

As exchange keys are 2048 bit and can thus encrypt only up to 190 bits of data, the [unsigned varint](https://github.com/multiformats/unsigned-varint) prefix MUST be smaller than $190 - 32 \cdot 2 = 126$ bytes.

In the current specification, valid prefixes always encode into 8 bytes. This means *decrypted* share payloads will always be 72 bytes long.

Share payload *ciphertexts* will always be 256 bytes long, due to the way RSAES-OAEP works.

### 3.2.1 Unsigned Varint Prefix

> WNFS (currently) uses the [multicodec private use range](https://github.com/multiformats/multicodec/#private-use-area) for prefixing and identifying share pointers.

Reviewer 2: I see this in the repo, but is this range canonical in a spec somewhere?

Author: I'm pretty sure that repository is the spec, the source of truth. It kind of sucks.
There used to be some effort in getting an RFC into IETF: https://github.com/multiformats/specs
But that seems to have faded out 2 years ago. Also, the [multiformats website](https://multiformats.io/) just links the repository.

> WNFS (currently) uses the [multicodec private use range](https://github.com/multiformats/multicodec/#private-use-area) for prefixing and identifying share pointers.

Reviewer 3: Awesome 👍 

Not for _right now,_ but these kinds of decryption pointers have come up in a couple contexts. if we haven't already, it may be worth us writing a standard for submission to multiformats.

Author: Yes. I want to play with them a bit, see where else they end up useful and plan to do exactly that when it feels useful.

| Prefix | Meaning | Decoded Unsigned Varint | Prefix Computation |
|-------:|---------|-------------------------|:-------------------|
| `0xceaec101` | Temporal Share Pointer | `0x30574e` | `varint_enc(0x300000 \| ascii_enc("WN"))` |
| `0xcfaec101` | Snapshot Share Pointer | `0x30574d` | `varint_enc(0x300000 \| ascii_enc("WN") + 1)` |

The `varint_enc` function encodes an [unsigned varint](https://github.com/multiformats/unsigned-varint).

### 3.2.2. Temporal Share Pointer

After the unsigned varint prefix `0xceaec101`, the temporal share pointer consists of
- a 32 byte private forest label used to look up the private node to decrypt in the private forest

> - a 32 byte [revision key](/spec/private-wnfs.md#3161-revision-key) to decrypt the private node and its temporal header.

Reviewer 3: 👍

### 3.2.3 Snapshot Share Pointer

> The snapshot share pointer works exactly like the temporal share pointer, except it is prefixed with `0xcfaec101` and instead of encoding a revision key, it encodes a [content key](/spec/private-wnfs.md#3162-content-key).

Reviewer 3: Purely out of curiosity: is there a pun or some other meaning behind `0xcfaec101`?

Reviewer 3: Nevermind, it's above... this is what I get for bouncing between tasks 😅

> This content key only gives access to a single revision, in case a user wanted to only share a single revision with someone else.

Reviewer 3: Interesting. Do we have a similar convention for all decryption pointers?

Author: Assuming you're referring to the unsigned varint prefixing, that's only for shared private files.
Signaling generally works differently for the rest of WNFS: We're using `type` keys in CBOR objects.
I initially wanted to do the same in this spec. However, I have to save bytes as I can only encrypt up to 190 bytes using RSA-OAEP. Using the unsigned varint and a packed byte format saves some bytes.

Author: Perhaps that's premature optimization. CBOR is fairly concise, too. A CBOR encoding would fit within <110 bytes.

Reviewer 3: From our call: this should be consistent with the private partition, whether that's by file name or multicodec prefix.

# 4 Share Lookup and Discovery

It's not assumed that recipients are notified over some secondary channel when they receive new shares.

> Thus it is possible to scan other user's file systems for files that were shared with a given exchange key. To do this, generate share labels as described in [section Share Label](#21-share-label) starting from 0 or the last counter used for lookup until the first missing label in the private forest is hit.

Reviewer 2: share label anchor?
